### PR TITLE
HD-1616. Expose a REST API on the ResourceManager to allow updating t…

### PR DESCRIPTION
**Pull request by raviprak-altiscale**
…he resources available to a NodeManager

This is following a very similar pattern to YARN-1702.

### CODE REVIEW

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site/src/site/markdown/ResourceManagerRest.md [hunk 1]
```diff
@@ -1753,6 +1753,70 @@ Response Body:
...
+
+### HTTP Operations Supported
+
+      * PUT
+
+### Query Parameters Supported
+
+      None
+
+### Elements of the NodeState object
```
Comment (ovy-altiscale) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site/src/site/markdown/ResourceManagerRest.md]: When reading this, I am not sure what NodeState is. Do you mean NodeCapacity?  Can you provide a link and/or make clear that the parameter is a json containing a "vcores" ?

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/webapp/RMWebServices.java [hunk 2]
```diff
@@ -734,7 +739,80 @@ public Response updateAppState(AppState targetState,
...
+    }
+
+    //TODO(raviprak) : See if you want to change memory too
+    int memory = node.getTotalCapability().getMemory();
+    int vCores = targetCapacity.getVCores();
+    Resource resource = Resource.newInstance(memory, vCores);
+    ResourceOption resourceOption = ResourceOption.newInstance(resource, -1);
+    Map<NodeId, ResourceOption> map = new HashMap<>();
+    map.put(node.getNodeID(), resourceOption);
+    final UpdateNodeResourceRequest req =
```
Comment (ovy-altiscale) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/webapp/RMWebServices.java:787]: FYI not sure if you wanted to put this in 2.4.1, but the client side has a bug in that version that makes the PB building crash always.   See https://github.com/VertiCloud/alti_node_decomissioner/blob/master/src/main/java/com/altiscale/decommissioner/Decommissioner.java#L240 

It's fixed somewhere before 2.7.1 though.

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/webapp/RMWebServices.java [hunk 3]
```diff
@@ -734,7 +739,80 @@ public Response updateAppState(AppState targetState,
 
     return Response.status(Status.OK).entity(ret).build();
   }
-  
+
+  // Copy-paste job from updateAppQueue() above . Almost exactly the same pattern as YARN-1702.
```
Comment (ovy-altiscale) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/webapp/RMWebServices.java]: Indeed, it looks very similar. Can some of this security dance be factored into a common method?

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site/src/site/markdown/ResourceManagerRest.md (hunk 1 above)
Comment (raviprak-altiscale) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site/src/site/markdown/ResourceManagerRest.md]: Yes! You're right! It's NodeCapacity. Will change.

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/webapp/RMWebServices.java (hunk 2 above)
Comment (raviprak-altiscale) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/webapp/RMWebServices.java:787]: Yeah! I had noticed your comment when I was reading the Decomissioner code. I suspect we will be free of Hadoop-2.4.1 shortly. We're down to almost 20%

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/webapp/RMWebServices.java (hunk 3 above)
Comment (raviprak-altiscale) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/webapp/RMWebServices.java]: I don't intend to contribute this upstream. To get this into apache, they'll likely ask for a much bigger effort to get a lot more RMAdmin Commands exposed via REST. This way, I'm isolating my code from modifications that may happen upstream, so I see value in replicating some code.

Comment (ovy-altiscale) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/webapp/RMWebServices.java]: Maybe put a TODO: refactor if merging upstream.

**Comment by ovy-altiscale** [comment]
LGTM

OK to self-merge after the README change.

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/webapp/RMWebServices.java (hunk 3 above)
Comment (raviprak-altiscale) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/webapp/RMWebServices.java]: Done

**Pull request merged**